client/src/utils/urlify.ts

```diff
@@ -1,21 +1,580 @@
+  ს: 's',
+  ტ: 't',
+  უ: 'u',
+  ფ: 'f',
+  ქ: 'q',
+  ღ: 'g',
+  ყ: 'y',
+  შ: 'sh',
+  ჩ: 'ch',
+  ც: 'c',
+  ძ: 'dz',
+  წ: 'w',
+  ჭ: 'ch',
+  ხ: 'x',
+  ჯ: 'j',
+  ჰ: 'h',
+};
+
+const ALL_DOWNCODE_MAPS = [
+  LATIN_MAP,
+  LATIN_SYMBOLS_MAP,
+  GREEK_MAP,
+  TURKISH_MAP,
+  ROMANIAN_MAP,
+  RUSSIAN_MAP,
+  UKRAINIAN_MAP,
+  CZECH_MAP,
+  SLOVAK_MAP,
+  POLISH_MAP,
+  LATVIAN_MAP,
+  ARABIC_MAP,
+  LITHUANIAN_MAP,
+  SERBIAN_MAP,
+  AZERBAIJANI_MAP,
+  GEORGIAN_MAP,
+];
+
+class Downcoder {
+  map: Record<string, string>;
+  regex: RegExp;
+
+  constructor() {
+    const map = {};
+
+    // eslint-disable-next-line no-plusplus
+    for (let i = 0; i < ALL_DOWNCODE_MAPS.length; i++) {
+      const lookup = ALL_DOWNCODE_MAPS[i];
+      for (const c in lookup) {
+        if (lookup.hasOwnProperty(c)) {
+          map[c] = lookup[c];
+        }
+      }
+    }
+
+    const chars: string[] = [];
+
+    for (const k in map) {
+      if (map.hasOwnProperty(k)) {
+        chars.push(k);
+      }
+    }
+
+    this.map = map;
+    this.regex = new RegExp(chars.join('|'), 'g');
+  }
+
+  downcode(s: string) {
+    return s.replace(this.regex, (m) => this.map[m]);
+  }
+}
+
+const downcoder = new Downcoder();
+
+/**
+ * IMPORTANT
+ * This util and the mapping is a direct port of Django's urlify.js util, without the need for a full Regex implementation.
+ * @see https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/urlify.js
+ */
+function URLify(str, numChars, allowUnicode = false) {
+  let s = str;
+  // changes, e.g., "Petty theft" to "petty-theft"
+  if (!allowUnicode) {
+    s = downcoder.downcode(s);
+  }
+  s = s.toLowerCase(); // convert to lowercase
+  // if downcode doesn't hit, the char will be stripped here
+  if (allowUnicode) {
+    // Keep Unicode letters including both lowercase and uppercase
+    // characters, whitespace, and dash; remove other characters.
+    // eslint-disable-next-line prefer-regex-literals
+    s = s.replace(s, RegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
+  } else {
+    s = s.replace(/[^-\w\s]/g, ''); // remove unneeded chars
   }
+  s = s.replace(/^\s+|\s+$/g, ''); // trim leading/trailing spaces
+  s = s.replace(/[-\s]+/g, '-'); // convert spaces to hyphens
+  s = s.substring(0, numChars); // trim to first num_chars chars
+  s = s.replace(/-+$/g, ''); // trim any trailing hyphens
+  return s;
 }
 
 /**
  * Returns the supplied string as a slug suitable for a URL using the vendor URLify util.
  * If the vendor util returns an empty string it will fall back to the slugify method.
  */
 export const urlify = (value: string, options = {}) => {
-  // URLify performs extra processing on the string (e.g. removing stopwords) and is more suitable
+  // URLify performs extra processing on the string and is more suitable
   // for creating a slug from the title, rather than sanitising a slug entered manually
-  const cleaned = window.URLify(value, 255);
+  const cleaned = URLify(value, 255);
```

Comment:
I'm surprised we never pass the options' `unicodeSlugsEnabled` value to `URLify` here. Doesn't this mean the `WAGTAIL_ALLOW_UNICODE_SLUGS` option require you to manually edit the slug to use unicode, and that the autogenerated one (from the title) will always use ASCII? Was that intentional?

Edit: I'm not sure why/how this was changed, but, compare the following:

- https://static-wagtail-v3-0.netlify.app/admin/pages/add/base/standardpage/60/
- https://static-wagtail-v4-0.netlify.app/admin/pages/add/base/standardpage/60/

In v3.0, if you enter e.g. "Spam & ıçüş" in the title field, the slug is generated as "spam-ıçüş". In v4.0, it becomes "spam-icus". I couldn't find whether it was intentional or not, will try to dig more...

P.s. I know that historically this used to be configured via `window.unicodeSlugsEnabled` in `_editor_js.html`, and we did away with in #10189. However, that happened in 5.0, so this issue has existed since before then.

Reply:
Yeah, I ran into a similar confusion a long time ago when doing some earlier work (even before Stimulus on this).

I do think we should review this current behaviour but for now I will leave as is and we can raise a new bug. Any changes to the output here could cause issues for pages that are NOT published with existing slugs. This is because it will do a comparison on load of the page to check if it should keep the title/slug in sync.

Reply:
OK it looks like it was me sadly ->  510a6bf350e4124193e63b076398010a94e019d8 #8748 (Released in Wagtail 4.0).

The original code had.

```js
const cleaned = URLify(val, 255, window.unicodeSlugsEnabled);
```

But when moved to `client/src/utils/text.ts`, it became

```js
    const cleaned = window.URLify(val, 255);
```

I will raise a separate bug for this.

Reply:
Bug raised https://github.com/wagtail/wagtail/issues/11828